File: chess.scala/Figure.scala
package chess

/**
  * Represents chess figure types.
  */
abstract sealed class FigureType
case object King extends FigureType
case object Queen extends FigureType
case object Rook extends FigureType
case object Bishop extends FigureType
case object Knight extends FigureType
case object Pawn extends FigureType

/**
  * Represents a figure, which has a type and a color.
  */
case class Figure(figureType: FigureType, figureColor: Color) {

  /**
    * Returns a one-character string representing the figure.
    */
  override def toString = (figureType, figureColor) match {
    case (King,   White) => "k"
    case (Queen,  White) => "q"
    case (Rook,   White) => "r"
    case (Bishop, White) => "b"
    case (Knight, White) => "n"
    case (Pawn,   White) => "p"
    case (King,   Black) => "K"
    case (Queen,  Black) => "Q"
    case (Rook,   Black) => "R"
    case (Bishop, Black) => "B"
    case (Knight, Black) => "N"
    case (Pawn,   Black) => "P"
  }

  /**
    * Returns a one-character string with the Unicode symbol representing the figure.
    */
  def figureSymbol = (figureType, figureColor) match {
    case (King,   White) => "\u2654"
    case (Queen,  White) => "\u2655"
    case (Rook,   White) => "\u2656"
    case (Bishop, White) => "\u2657"
    case (Knight, White) => "\u2658"
    case (Pawn,   White) => "\u2659"
    case (King,   Black) => "\u265a"
    case (Queen,  Black) => "\u265b"
    case (Rook,   Black) => "\u265c"
    case (Bishop, Black) => "\u265d"
    case (Knight, Black) => "\u265e"
    case (Pawn,   Black) => "\u265f"
  }

}

/*
scaladoc chess\*.scala
scalac chess\*.scala
scala
import chess._
Figure(King, White) // k
Figure(King, Black) // K
Figure(Knight, White) // n
Figure(King, White).figureSymbol
Figure(King, Black).figureSymbol
Figure(Knight, White).figureSymbol
:q
*/

File: chess.py/figure.py
from dataclasses import dataclass
from enum import Enum, auto

from color import Color


class FigureType(Enum):
    """
    Represents chess figure types.
    """

    King = auto()
    Queen = auto()
    Rook = auto()
    Bishop = auto()
    Knight = auto()
    Pawn = auto()


@dataclass
class Figure:
    """
    Represents a figure, which has a type and a color.
    """

    figure_type: FigureType
    figure_color: Color

    def __str__(self):
        """
        Returns a one-character string representing the figure.

        >>> print(Figure(FigureType.King, Color.White))
        k

        >>> print(Figure(FigureType.King, Color.Black))
        K
        """
        match self:
            case Figure(FigureType.King, Color.White):
                return "k"
            case Figure(FigureType.Queen, Color.White):
                return "q"
            case Figure(FigureType.Rook, Color.White):
                return "r"
            case Figure(FigureType.Bishop, Color.White):
                return "b"
            case Figure(FigureType.Knight, Color.White):
                return "n"
            case Figure(FigureType.Pawn, Color.White):
                return "p"
            case Figure(FigureType.King, Color.Black):
                return "K"
            case Figure(FigureType.Queen, Color.Black):
                return "Q"
            case Figure(FigureType.Rook, Color.Black):
                return "R"
            case Figure(FigureType.Bishop, Color.Black):
                return "B"
            case Figure(FigureType.Knight, Color.Black):
                return "N"
            case Figure(FigureType.Pawn, Color.Black):
                return "P"

    def figure_symbol(self):
        """
        Returns a unicode symbol representing the figure.
        """
        match self:
            case Figure(FigureType.King, Color.White):
                return "\u2654"
            case Figure(FigureType.Queen, Color.White):
                return "\u2655"
            case Figure(FigureType.Rook, Color.White):
                return "\u2656"
            case Figure(FigureType.Bishop, Color.White):
                return "\u2657"
            case Figure(FigureType.Knight, Color.White):
                return "\u2658"
            case Figure(FigureType.Pawn, Color.White):
                return "\u2659"
            case Figure(FigureType.King, Color.Black):
                return "\u265a"
            case Figure(FigureType.Queen, Color.Black):
                return "\u265b"
            case Figure(FigureType.Rook, Color.Black):
                return "\u265c"
            case Figure(FigureType.Bishop, Color.Black):
                return "\u265d"
            case Figure(FigureType.Knight, Color.Black):
                return "\u265e"
            case Figure(FigureType.Pawn, Color.Black):
                return "\u265f"
